mr/src/main/java/org/apache/iceberg/mr/mapred/IcebergInputFormat.java

```diff
@@ -0,0 +1,259 @@
+    for (int i = 0; i < tasks.size(); i++) {
+      splits[i] = new IcebergSplit(tasks.get(i), location);
+    }
+    return splits;
+  }
+
+  @Override
+  public RecordReader getRecordReader(InputSplit split, JobConf job, Reporter reporter) throws IOException {
+    return new IcebergRecordReader(split, job);
+  }
+
+  @Override
+  public boolean shouldSkipCombine(Path path, Configuration conf) throws IOException {
+    return true;
+  }
+
+  public class IcebergRecordReader implements RecordReader<Void, IcebergWritable> {
+    private JobConf conf;
+    private IcebergSplit split;
+
+    private Iterator<FileScanTask> tasks;
+    private CloseableIterable<Record> reader;
+    private Iterator<Record> recordIterator;
+    private Record currentRecord;
+    private boolean reuseContainers;
+
+    public IcebergRecordReader(InputSplit split, JobConf conf) throws IOException {
+      this.split = (IcebergSplit) split;
+      this.conf = conf;
+      this.reuseContainers = conf.getBoolean(REUSE_CONTAINERS, false);
+      initialise();
+    }
+
+    private void initialise() {
+      tasks = split.getTask().files().iterator();
+      nextTask();
+    }
+
+    private void nextTask() {
+      FileScanTask currentTask = tasks.next();
+      DataFile file = currentTask.file();
+      InputFile inputFile = HadoopInputFile.fromLocation(file.path(), conf);
+      Schema tableSchema = table.schema();
+
+      reader = IcebergReaderFactory.createReader(file, currentTask, inputFile, tableSchema, reuseContainers);
+      recordIterator = reader.iterator();
+    }
+
+    @Override
+    public boolean next(Void key, IcebergWritable value) {
+      if (recordIterator.hasNext()) {
+        currentRecord = recordIterator.next();
+        value.setRecord(currentRecord);
+        return true;
+      }
+
+      if (tasks.hasNext()) {
+        try {
+          reader.close();
+        } catch (IOException e) {
+          LOG.error("Error closing reader", e);
+        }
+        nextTask();
+        currentRecord = recordIterator.next();
+        value.setRecord(currentRecord);
+        return true;
+      }
+      return false;
+    }
+
+    @Override
+    public Void createKey() {
+      return null;
+    }
+
+    @Override
+    public IcebergWritable createValue() {
+      IcebergWritable record = new IcebergWritable();
+      record.setRecord(currentRecord);
+      record.setSchema(table.schema());
+      return record;
+    }
+
+    @Override
+    public long getPos() throws IOException {
+      return 0;
+    }
+
+    @Override
+    public void close() throws IOException {
+      reader.close();
+    }
+
+    @Override
+    public float getProgress() throws IOException {
+      return 0;
+    }
+  }
+
+  /**
+   * FileSplit is extended rather than implementing the InputSplit interface due to Hive's HiveInputFormat
+   * expecting a split which is an instance of FileSplit.
+   */
+  private static class IcebergSplit extends FileSplit {
```

Comment:
Can `IcebergSplit` extend `v2InputSplit` and implement `v1InputSplit`
I tried it and Intellij does not complain. Seems like it would compile
```
static class IcebergSplit extends InputSplit implements Writable, org.apache.hadoop.mapred.InputSplit {..
```
If this works, maybe we can have a common InputSplit for both Mrv1 and Mrv2

Also since RecordReader is the most complex bit. I think we should we able to reuse it for Mrv1 RecordReader.
What if v1Record internally kept a reference to v2RecordReader? .  Seems like we would only need to call `initialize` correctly. For the we can use `org.apache.hadoop.mapred.TaskAttemptContextImpl#TaskAttemptContextImpl(org.apache.hadoop.mapred.JobConf, org.apache.hadoop.mapred.TaskAttemptID)`.   For the `InputSplit` parameter, I think we might need to tweak  `mapreduce.IcebergRecordReader` a little, like so

```
@Override
    public void initialize(InputSplit split, TaskAttemptContext newContext) {
      initialize((IcebergSplit) split, newContext);
    }

    public void initialize(IcebergSplit split, TaskAttemptContext newContext) {
```

This will work since `IcebergSplit1 extends both v1 and v2 InputSplit

Reply:
This can potentially help us rely delegate all calls to `mapreduce.IcebergRecordReader`

Reply:
I'll give this a go and see if I can get it to work.

Reply:
Reusing as much as we can would be ideal.

Reply:
From our integration tests, what actually happens in Hive is that it wraps the InputFormat with a HiveInputFomat which in turn does an `instanceof FileSplit` which is why we had to extend `FileSplit` here otherwise it throws a RuntimeException and fails. See https://github.com/apache/hive/blob/rel/release-2.3.7/ql/src/java/org/apache/hadoop/hive/ql/io/HiveInputFormat.java#L134. So I don't think we can have a single InputSplit class. I will look into if there is some other way to share as much of the RecordReader code as possible.